packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,484 @@
+      print(chalk.red(`Error deleting objects: ${err}`));
+    }
+  }
+
+  private async bootstrapBucketName(sdk: ISDK, bootstrapStackName: string): Promise<string> {
+    const info = await ToolkitInfo.lookup(this.props.resolvedEnvironment, sdk, bootstrapStackName);
+    return info.bucketName;
+  }
+
+  private async bootstrapQualifier(sdk: ISDK, bootstrapStackName: string): Promise<string | undefined> {
+    const info = await ToolkitInfo.lookup(this.props.resolvedEnvironment, sdk, bootstrapStackName);
+    return info.bootstrapStack.parameters.Qualifier;
+  }
+
+  private async numObjectsInBucket(s3: S3, bucket: string): Promise<number> {
+    const response = await s3.listObjectsV2({ Bucket: bucket }).promise();
+    return response.KeyCount ?? 0;
+  }
+
+  /**
+   * Generator function that reads objects from the S3 Bucket in batches.
+   */
+  private async *readBucketInBatches(s3: S3, bucket: string, batchSize: number = 1000): AsyncGenerator<S3Asset[]> {
+    let continuationToken: string | undefined;
+
+    do {
+      const batch: S3Asset[] = [];
+
+      while (batch.length < batchSize) {
+        const response = await s3.listObjectsV2({
+          Bucket: bucket,
+          ContinuationToken: continuationToken,
+        }).promise();
+
+        response.Contents?.forEach((obj) => {
+          const key = obj.Key ?? '';
+          const size = obj.Size ?? 0;
+          if (obj.Key) {
+            batch.push(new S3Asset(bucket, key, size));
+          }
+        });
+
+        continuationToken = response.NextContinuationToken;
+
+        if (!continuationToken) break; // No more objects to fetch
+      }
+
+      if (batch.length > 0) {
+        yield batch;
+      }
+    } while (continuationToken);
+  }
+
+  /**
+   * Fetches all relevant stack templates from CloudFormation. It ignores the following stacks:
+   * - stacks in DELETE_COMPLETE or DELET_IN_PROGRES stage
+   * - stacks that are using a different bootstrap qualifier
+   *
+   * It fails on the following stacks because we cannot get the template and therefore have an imcomplete
+   * understanding of what assets are being used.
+   * - stacks in REVIEW_IN_PROGRESS stage
+   */
+  private async fetchAllStackTemplates(cfn: CloudFormation, qualifier?: string) {
+    const stackNames: string[] = [];
+    await paginateSdkCall(async (nextToken) => {
+      let response = await cfn.listStacks({ NextToken: nextToken }).promise();
+
+      // We cannot operate on REVIEW_IN_PROGRESS stacks because we do not know what the template looks like in this case
+      // If we encounter this status, we will wait up to a minute for it to land before erroring out.
+      const reviewInProgressStacks = response.StackSummaries?.filter(s => s.StackStatus === 'REVIEW_IN_PROGRESS') ?? [];
+      if (reviewInProgressStacks.length > 0) {
+        const updatedRequest = await this.waitForStacksAndRefetch(cfn, response, reviewInProgressStacks);
+        response = await updatedRequest.promise();
```

Comment:
The shape of the code is confusing me a little. Why not a `while` loop here?

It seems this is what we want:

```ts
let stacks;
do {
  stacks = listAllStacks();
} while (stacks.some(isReviewInProgress));

// Now we can use 'stacks'
```

In practice, I don't like post-test loops (`do/while` as opposed to `while/do`, a pre-test loop) all that much. And we also need to account for sleeping. So in practice, it might be:

```ts
let stacks;

let sleepMs = 500;
while (true) {
  stacks = listAllStacks();
  if (!stacks.some(isReviewInProgress)) {  
    break;
  }
  await sleep(Math.floor(Math.random() * sleepMs));
  sleepMs = Math.min(sleepMs * 2, 60_000);
}

// Use stacks
```

Or even better, put the pre-test loop in a function and use `return` instead of shared variable + `break`:

```ts
const stacks = await listStacksNotBeingReviewed();
```